src/libponyc/pass/verify.c

```diff
@@ -90,19 +96,386 @@ static bool verify_assign_lvalue(pass_opt_t* opt, ast_t* ast)
   return true;
 }
 
+static size_t funref_hash(ast_t** key)
+{
+  return ponyint_hash_ptr(*key);
+}
+
+static bool funref_cmp(ast_t** a, ast_t** b)
+{
+  return *a == *b;
+}
+
+DECLARE_HASHMAP(consume_funs, consume_funs_t, ast_t*);
+
+DEFINE_HASHMAP(consume_funs, consume_funs_t, ast_t*, funref_hash,
+  funref_cmp, NULL);
+
+static const char* get_multi_ref_name(ast_t* ast)
```

Comment:
Same comment here as I gave for `generate_multi_dot_name` - I'd like to have a comment here describing what, why, and an example or two.

In fact I'm even more curious as to how this function is similar/different to `generate_multi_dot_name` in its behaviour and if the two could possibly share any common implementation.

Reply:
This function generates the fully qualified string (without the `this`) for a reference (i.e. `a.b.c.d.e`) and ensures it is part of the compiler `stringtab`. This is basically the same logic as the `generate_multi_dot_name` from `refer.c`. It would likely be possible to combine the two into a single function but I haven't taken the time to figure out exactly how easy/possible that would be. It is used to generate this fully qualified string for the field being confused to tracking its `consume`d status via the ast `symtab`. It is used to ensure that no parent (e.g. `a.b.c`) of a consumed field (e.g. `a.b.c.d.e`) is consumed.